devnet-sdk/testing/systest/systest.go

```diff
@@ -28,22 +87,40 @@ func SystemTest(t BasicT, f SystemTestFunc, validators ...PreconditionValidator)
 	for _, validator := range validators {
 		ctx, err := validator(wt, sys)
 		if err != nil {
-			t.Skipf("validator failed: %v", err)
+			h.handlePreconditionError(t, err)
```

Comment:
At a glance, I'm a bit unclear as to when to call `handlePreconditionError`?

Reply:
that's just an internal helper to dispatch precondition errors to Skip / Fatal depending on the desired behavior.
So anywhere the previous implementation was calling Skipf:
- while iterating over the validators  (here)
- while evaluating if the resulting system has the right features (below)